### ISSUE TITLE
Improve CommonFinders for Dart 2

### ISSUE THREAD

**Issue by munificent** [flutter/flutter#15833]
**TL;DR: CommonFinders should probably have some API changes to be a better Dart 2 API, but doing so without breaking Dart 1 support is hard.**

Say we have a class like:

```dart
class ValuePicker<T> extends Widget {
  T value;
}
```

And the application code is creating pickers of specific types like `ValuePicker<int>`, `ValuePicker<String>`, etc. Tests for the application might look like:

```dart
var picker = find.byType(ValuePicker).first;
```

This works in Dart 1, but does not and will not work in Dart 2. The implementation of `find.byType()` does something like:

```dart
widget.runtimeType == widgetType
```

In Dart 2, the Type literal `ValuePicker` gives you a type object representing `ValuePicker<dynamic>`. The runtime type of a `ValuePicker<int>` is `ValuePicker<int>`, which is *not* equivalent to `ValuePicker<dynamic>`. The above `.descendant()` call will fail to match any widgets.

You could imagine fixing this by just passing a generic type instead of the raw ValuePicker:

```
var picker = find.byType(ValuePicker<int>).first;
```

Alas, that's not valid Dart syntax (or at least not yet, and it won't be until after 2 if at all).

This hack *does* work in Dart 2:

```dart
Type reify<T>() => T;

var picker = find.byType(reify<ValuePicker<int>>()).first;
```

But that's pretty ugly. Better would be to make the methods themselves generic and have them effectively do the same thing:

```dart
var picker = find.byType<ValuePicker<int>>().first;
```

That looks nice to me and is, I think, what an idiomatic Dart 2 solution looks like.

Unfortunately, it (and the previous `reify<T>()` hack) do *not* work in Dart 1. :( Dart 1 doesn't reify generic method type arguments, so the above code would simply not do anything useful.

This would work, I think:

```dart
class Reify<T> {
  Type get type => T;
}

var picker = find.byType(new Reify<ValuePicker<int>>().type).first;
```

But that's pretty hideous. Simpler is:

```dart
var picker = find.byWidgetPredicate((widget) => widget is ValuePicker<int>).first;
```

That avoids clever generic stuff and does the right thing in Dart 1 and 2, but it's also really verbose. To get current code that needs to support both Dart 1 and 2, the latter fix is an OK workaround, but I'm filing this bug so you folks can think about what the desired API is for this after Dart 1 support is no longer needed.

My suggestion is to turn `byType()` and friends into generic methods:

```dart
Finder widgetWithText<T extends Widget>(String text, { bool skipOffstage: true });

Finder byType<T extends Widget>({ bool skipOffstage: true });

Finder byElementType<T extends Element>({ bool skipOffstage: true });
```

That gives you a few benefits:

* Compile errors if you, for example, try to call `byType()` with a type that isn't a widget.

* Implementations of these methods can use corelib APIs like `Iterable.whereType<T>()` which requires the filtered type be a type argument.

* As noted above, works seamlessly with types that are generic.

* Lifts the type such that it's visible to the type system and static analysis. For example, say you later add a method to combine `byType()` and `byWidgetPredicate()`:

    ```dart
    Finder byPredicate<T extends Widget>(bool Function(T) predicate);
    ```

    Note that the callback takes a `T` — the actual widget type being filtered — not a base Widget. That way, the body of the callback can access properties on the specific widget type.

**Comment by Hixie** [comment]
I think we should just fix Dart to allow `Foo<Bar>` to be used as a Type literal. It's not really viable to not have that IMHO. All the other options here have terrible ergonomics.

**Comment by Hixie** [comment]
It strikes me that we have several other APIs that take `Type`s, even in the core framework (as opposed to "just" the test framework), and I suspect they'll all be affected in the same way.

See for instance:
* `BuildContext.inheritFromWidgetOfExactType`
* `BuildContext.ancestorInheritedElementForWidgetOfExactType`
* `BuildContext.ancestorWidgetOfExactType`
* `RawGestureDetector.gestures`

The first one of these in particular is used in many applications and it would be very unfortunate to break that API.

One interesting example is `Localizations.resourcesFor`, which takes both a type argument and an argument of type `Type`, which we assert to be equal.

**Comment by Hixie** [comment]
The `RawGestureDetector.gestures` case is a `Map<Type, ...>`, which is a common construct in general (not necessarily in the Flutter framework specifically), which is broken by this limitation in Dart2.

**Comment by Hixie** [comment]
Alternatively, maybe we just break Dart1 compat with this change, and go straight to the idiomatic Dart2 solution of using actual type arguments.

**Comment by yjbanov** [comment]
I'm OK with breaking Dart 1 compat, and it seems we don't have to rush and break it right now, given that the `find.byWidgetPredicate` workaround exists. I think the following plan could work:

1. move usages of `byType` that need to match on type args to `find.byWidgetPredicate`
2. give users time to move to Dart 2
3. break Dart 1 compat by moving to `find.byType<TargetType>()`
4. users move to from `find.byWidgetPrediate` back to `find.byType`

**Comment by munificent** [comment]
> I think we should just fix Dart to allow `Foo<Bar>` to be used as a Type literal.

Yes, I believe the language team intends that, it just didn't make it into 2.0. I think it might get tangled in some corners of the grammar, though, which may be why it's not in and why it's hard to add.

For what my opinion is worth, I like @yjbanov's plan.

**Comment by yjbanov** [comment]
If the language supports `find.byType(ValuePicker<int>)` in the future, then we can change step #&#xfeff;3 to:

3. add support for `find.byType(ValuePicker<int>)`

All other steps would stay the same. Either way, since there's no rush, there's enough time to discuss which way to go.

**Comment by Hixie** [comment]
I'm unhappy about breaking users twice for no benefit, especially if the end result is no change on their part. I don't think our developers should put up with us yanking them around like that. If we're going to make a breaking change here, we should make a single breaking change, once, and with good reason.

"We didn't get around to adding this to the language" is not a good reason, nor is "we didn't bother to implement this in the old VM", and nor is "it's hard".

**Comment by a14n** [comment]
IIRC `Type` will/might become a generic class (`Type<T>`). Associated with _allowing `Foo<Bar>` to be used as a Type literal_ the current apis could remain almost the same from a user point of view.

```diff
-  InheritedWidget inheritFromWidgetOfExactType(Type targetType);
+  @optionalTypeArgs
+  T inheritFromWidgetOfExactType<T extends InheritedWidget>(Type<T> targetType);
```

It has the benefits:
- to improve typing:
  - the user has to provide a `targetType` that is a child of `InheritedWidget`
  - the return type is more accurate
  - no need of an implicite downcast on call site
- to keep the type mandatory (with `inheritFromWidgetOfExactType<T extends InheritedWidget>()` there's no error if you forget the generic type and call `context.inheritFromWidgetOfExactType()`)
- to limit breakage on existing code (if there's not bad parameter as `targetType` - btw I've found some errors in flutter codebase)

**Comment by markusaksli-nc** [comment]
Closing as Dart 2 has landed and this is no longer relevant, feel free to reopen if this is not the case.
